File: LeetCode/src/dp/CoinChange.java
package dp;

import java.util.Arrays;

public class CoinChange {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];

        Arrays.fill(dp, amount + 1);

        dp[0] = 0;

        for (int coin : coins) {
            for (int j = coin; j <= amount; j++) {
                if (j >= coin) {
                    dp[j] = Math.min(dp[j], dp[j - coin] + 1);
                }
            }
        }

        return (dp[amount] == Integer.MAX_VALUE) ? -1 : dp[amount];
    }

    public static void main(String[] args) {
        CoinChange solution = new CoinChange();
        int[] coins = {1, 2, 5};
        int amount = 11;
        System.out.println(solution.coinChange(coins, amount)); // Output: 3
    }
}

File: LeetCode_Python/dp/CoinChange.py
class CoinChange:
    def coinChange(self, coins: list[int], amount: int) -> int:
        dp = [amount + 1] * (amount + 1)
        dp[0] = 0

        for i in range(1, amount + 1):
            for coin in coins:
                if i >= coin:
                    dp[i] = min(dp[i], dp[i - coin] + 1)

        return dp[amount] if dp[amount] != amount + 1 else -1


# Time Complexity: O(n * m) where n is the amount and m is the number of coins
# Space Complexity: O(n) where n is the amount


if __name__ == "__main__":
    solution = CoinChange()

    print("Excepted Output: 3")
    print("Actual Output: ", solution.coinChange([1, 2, 5], 11))

    print("Excepted Output: -1")
    print("Actual Output: ", solution.coinChange([2], 3))
